pkg/services/sqlstore/migrator/migrator.go

```diff
@@ -87,7 +94,32 @@ func (mg *Migrator) GetMigrationLog() (map[string]MigrationLog, error) {
 	return logMap, nil
 }
 
-func (mg *Migrator) Start() error {
+func (mg *Migrator) Start(isDatabaseLockingEnabled bool, lockAttemptTimeout int) (err error) {
+	if !isDatabaseLockingEnabled {
+		return mg.run()
+	}
+
+	return mg.InTransaction(func(sess *xorm.Session) error {
+		mg.Logger.Info("Locking database")
+		if err := casRestoreOnErr(&mg.isLocked, false, true, ErrMigratorIsLocked, mg.Dialect.Lock, LockCfg{Session: sess, Timeout: lockAttemptTimeout}); err != nil {
+			mg.Logger.Error("Failed to lock database", "error", err)
+			return err
+		}
+
+		defer func() {
+			mg.Logger.Info("Unlocking database")
+			unlockErr := casRestoreOnErr(&mg.isLocked, true, false, ErrMigratorIsUnlocked, mg.Dialect.Unlock, LockCfg{Session: sess})
+			if unlockErr != nil {
+				mg.Logger.Error("Failed to unlock database", "error", unlockErr)
+			}
+		}()
+
+		// migration will run inside a nested transaction
```

Comment:
I don't think all the migrations will run in the same transaction https://github.com/grafana/grafana/pull/44101/files#diff-cc26a898455f75d00301a5d13594c160a6558b4152fcd5c47b502375a03d3675R224

But they don't have to either since the whole database is locked :)

Reply:
I'm not sure I understand why they won't run in the same transaction: we start a new transaction inside [migrator.run()](https://github.com/grafana/grafana/blob/cc5f372e2d50d486d4c372a2f2762e6a93e915d7/pkg/services/sqlstore/migrator/migrator.go#L150)